controllers/apply-next/form-router.js

```diff
@@ -81,9 +81,112 @@ function translateForm(formModel, locale, formData) {
 
     return clonedForm;
 }
+const FORM_STATES = {
+    incomplete: {
+        type: 'INCOMPLETE',
+        label: 'In progress'
+    },
+    complete: {
+        type: 'COMPLETE',
+        label: 'Complete'
+    },
+    invalid: {
+        type: 'INVALID',
+        label: 'Invalid'
+    },
+    empty: {
+        type: 'EMPTY',
+        label: 'Not started'
+    }
+};
+
+function validateFormState(form, formErrors) {
+    const clonedForm = cloneDeep(form);
+    const invalidFieldNames = formErrors.map(e => e.param);
+
+    clonedForm.sections.map(section => {
+        // if a step has any errors, return and mark section as incomplete
+        section.steps.forEach(step => {
+            step.fieldsets.map(fieldset => {
+                fieldset.fields.map(field => {
+                    if (invalidFieldNames.indexOf(field.name) !== -1) {
+                        // Is this field in the error array?
+                        // (eg. it failed validation)
+                        // @TODO should this just be marked as incomplete?
+                        field.state = FORM_STATES.invalid;
+                    } else {
+                        // Otherwise, it must be valid
+                        field.state = FORM_STATES.complete;
+                    }
+                    return field;
+                });
+                return fieldset;
+            });
+
+            // See if this step's fieldsets are all empty (as opposed to invalid)
+            const stepIsEmpty = isEmpty(
+                pickBy(
+                    step.fieldsets.reduce((acc, fieldset) => {
+                        return concat(fieldset.fields.map(f => f.value), acc);
+                    }, []),
+                    identity
+                )
```

Comment:
This looks a bit magic – it's similar to the code we use in the grants service to strip out missing/invalid fields (eg. `_.isEmpty(_.pickBy([undefined, null], _.identity))`). It basically reduces each step down to an array of the input values after removing invalid/missing ones, then checks to see if the resulting array is empty or not.